### PR TITLE
Improve point estimate extraction and related

### PR DESCRIPTION
**Pull request by xEnVrE**
Within this PR:

#### Filtering utilities
- Change ctor of class `EstimatesExtraction` in order to provide the size of the state, linear and circular
- Re-implement method `EstimatesExtraction::mean` in order to exploit the information on the linear and circular size of the state
- Change ctor of class `HistoryBuffer` in order to provide the size of the state
- Implement MAP (maximum a posteriori) estimate within method `EstimatesExtraction::map` taken from: 
    - Saha, S., Boers, Y., Driessen, H., Mandal, P. K., Bagchi, A. (2009), Particle Based MAP State Estimation: A Comparison, 12th International Conference on Information Fusion, Seattle, WA, USA, July 6-9, 2009, https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=5203651
- Implement a variant of the method `EstimatesExtraction::extract` that takes extra arguments required to extract the MAP estimate, namely the particle weights relative to the previous filtering step, the likelihoods relative to the current step and a matrix containing the Markov transition probabilities between all the particles at the previous filtering step and all the particles at the current filtering step
- The methods `EstimatesExtraction::extract` return a `std::pair` comprising a boolean, indicating if the extracted estimate is valid, and the extracted estimate. This is required since the user may request an extraction method without providing all the information required for it (e.g. the user asks for the map estimate but only provides particles and current weights)
- Re-implement class `EstimatesExtraction` so that particle weights are used directly in the log space (this is of great help when MAP and mode need to be extracted since the index of the best particle is searched among more well scaled, as compared to the linear space, values)

#### Filtering functions
- Add pure virtual method `GaussianPrediction::getStateModel` since it is required to implement properly `GPFPrediction::getStateModel` that in turn may be required to implement MAP estimate extraction within Gaussian particle filters
- Implement `KFPrediction::getStateModel`
- Implement `UKFPrediction::getStateModel`
- Properly implement `GPFPrediction::getStateModel`
- Properly implement `GPFCorrection::getLikelihood`

#### Fixes
- Fix missing `const`s in signature of methods `StateModel::getTransitionProbability` and `WhiteNoiseAcceleration::getTransitionProbability`

#### Test
- Use class `EstimatesExtraction` to extract the mean in `testUPF`
- Add extraction/logging of mean in `testSIS` to facilitate inspection of the algorithm outcome
- Reduce number of particles in `testUPF` to avoid too large computations time in `Debug`.
- Add test `testUPF_MAP` testing MAP estimate extraction within a UPF particle filter

#### Changelog
- Update `CHANGELOG.md`

#### CMake
- Update version to `0.9.100` since API of class `EstimatesExtraction` is broken

### CODE REVIEW

File: src/BayesFilters/src/EstimatesExtraction.cpp [hunk 1]
```diff
@@ -235,11 +236,20 @@ VectorXd EstimatesExtraction::map
...
+     */
 
     ArrayXd::Index map_index;
 
-    ((transition_probabilities * previous_weights).array() * likelihoods.array()).maxCoeff(&map_index);
+    VectorXd values(particles.cols());
+
+    double eps = std::numeric_limits<double>::min();
+    for (std::size_t i = 0; i < values.size(); i++)
+        values(i) = log(likelihoods(i) + eps) + utils::log_sum_exp((transition_probabilities.row(i).array() + eps).log() + previous_weights.array());
```
Comment (claudiofantacci) [inline src/BayesFilters/src/EstimatesExtraction.cpp]: I don't really like to pass eps to logsumexp as the only point of failure would be the log evaluated in 0.
To be precise, a good apporach would be to have inside logsumexp an eps summed only in the evaluation of the log.
Anyway, let's leave it like this and keep in mind that we may change it in some future commits.

**Review by claudiofantacci — DISMISSED** [review]
This is amazing! Thanks for the PR!
LGTM 🎉

**Comment by claudiofantacci** [comment]
@xEnVrE, before merging, please have a look at the build errors in travis.

File: src/BayesFilters/src/EstimatesExtraction.cpp (hunk 1 above)
Comment (xEnVrE) [reply src/BayesFilters/src/EstimatesExtraction.cpp]: You are right, I agree on changing it.

**Comment by xEnVrE** [comment]
Fixed the issue occured while testing `testUPF_MAP` in `Debug`.

I also reduced the number of particles in `testUPF` and `testUPF_MAP` to avoid too large computations time while testing in `Debug`.

Finally, there were some missing `override` keywords in `UKFPrediction::getStateModel`.

**Review by claudiofantacci — APPROVED** [review]
I'll merge after tests pass.

**Pull request merged**